### PR TITLE
Add vLLM grouped gemm backend for MoE inference

### PR DESCRIPTION
**Pull request by santhnm2**
# What does this PR do ?
Ports the vLLM grouped gemm kernel to the inference-optimized MoE backend.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/inference/moe/permute.py [hunk 1]
```diff
@@ -40,27 +51,30 @@ def _count_local_tokens_kernel(
...
+    block_start = pid * blocks_per_cta
+    block_end = tl.minimum(block_start + blocks_per_cta, total_blocks)
+
+    for block_id in tl.range(block_start, block_end):
+        offsets = block_id * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+        mask = offsets < valid_pairs
+        expert_ids = tl.load(routing_map_ptr + offsets, mask=mask, other=-1)
+        local_ids = expert_ids - local_expert_start
+        is_local = (local_ids >= 0) & (local_ids < num_local_experts) & mask
+        tl.atomic_add(tokens_per_expert_ptr + local_ids, 1, mask=is_local)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/moe/permute.py]: Does this reduce the number of global-mem atomic adds by `BLOCK_SIZE`? Does this also benefit the torch grouped gemm backend?

File: megatron/core/inference/moe/permute.py [hunk 2]
```diff
@@ -40,27 +51,30 @@ def _count_local_tokens_kernel(
...
-    is_local = (local_ids >= 0) & (local_ids < num_local_experts) & mask
-    tl.atomic_add(tokens_per_expert_ptr + local_ids, 1, mask=is_local)
+
+    total_blocks = tl.cdiv(valid_pairs, BLOCK_SIZE)
+    blocks_per_cta = tl.cdiv(total_blocks, num_sms)
+    block_start = pid * blocks_per_cta
+    block_end = tl.minimum(block_start + blocks_per_cta, total_blocks)
+
+    for block_id in tl.range(block_start, block_end):
+        offsets = block_id * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/moe/permute.py]: Should we guard the entirety of this loop under `if block_start < valid_pairs:`?

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

File: megatron/core/inference/moe/vllm_fused_moe.py [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+                a_ptrs,
+                mask=token_mask[:, None] & (offs_k[None, :] < K - k * BLOCK_SIZE_K),
+                other=0.0,
+            )
+            b = tl.load(b_block_ptr, boundary_check=(0, 1), padding_option="zero")
+            accumulator += tl.dot(a, b)
+            a_ptrs += BLOCK_SIZE_K * stride_ak
+            b_block_ptr = tl.advance(b_block_ptr, (BLOCK_SIZE_K, 0))
+
+        if FUSE_SQUARED_RELU:
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/moe/vllm_fused_moe.py:232]: If we are fusing squared relu into this, could we also try fusing the second grouped gemm?

File: megatron/core/inference/moe/vllm_fused_moe.py [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+    sorted_token_ids_ptr,
+    expert_ids_ptr,
+    max_sorted,
+    max_blocks,
+    SENTINEL: tl.constexpr,
+    BLOCK: tl.constexpr,
+):
+    """Initialize sorted_token_ids to SENTINEL and expert_ids to -1."""
+    pid = tl.program_id(0)
+    offs = pid * BLOCK + tl.arange(0, BLOCK)
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/moe/vllm_fused_moe.py]: Same comment to guard the rest of the code under `if pid * BLOCK < max_sorted`

File: megatron/core/transformer/moe/moe_layer.py [hunk 5]
```diff
@@ -340,6 +340,14 @@ def __init__(
                     "inference_grouped_gemm_backend='torch' requires "
                     "torch.nn.functional.grouped_mm (available since PyTorch 2.10)."
                 )
+            elif config.inference_grouped_gemm_backend == 'vllm':
+                try:
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/moe/moe_layer.py]: is this the right place to check imports? Could we do this at the top of this file?

File: megatron/core/transformer/moe/token_dispatcher_inference.py [hunk 6]
```diff
@@ -58,6 +58,11 @@ class InferenceAllGatherDispatcherBase(MoEAllGatherTokenDispatcher):
     # so that experts.py can always call _valid_tokens() on this base class.
     _valid_tokens_tensor: Optional[torch.Tensor] = None
 
+    # Host-side mirror of the total valid token count. Set each step in
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/moe/token_dispatcher_inference.py]: This is technically not a mirror of `_valid_tokens_tensor`. `_valid_tokens_tensor` contains the sum of token counts across all ranks.` _host_valid_tokens` is a reasonable estimate of this sum. We should update the variable name and the comment to reflect this.

**Review by sidsingh-nvidia — DISMISSED** [review]

File: megatron/core/inference/moe/vllm_fused_moe.py (hunk 3 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/moe/vllm_fused_moe.py:232]: on second thought this might be more complicated. BLOCK_K and BLOCK_N need to swap roles and that might be tricky.

File: megatron/core/inference/moe/permute.py (hunk 1 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/moe/permute.py]: Could we call tl.histogram on local_ids and then atomically add the histogram? I think that should optimize the number of global mem atomic adds.

Comment (santhnm2) [reply megatron/core/inference/moe/permute.py]: I don't think it reduces the number of atomic adds, just makes the grid setup have better CTA utilization. It should benefit the torch grouped gemm backend as well.

I tried to implement the `tl.histogram` optimization but ran into some triton compiler bug, will add a TODO to revisit this later.

File: megatron/core/inference/moe/permute.py (hunk 2 above)
Comment (santhnm2) [reply megatron/core/inference/moe/permute.py]: I added a `if block_start < total_blocks` guard

File: megatron/core/inference/moe/vllm_fused_moe.py (hunk 4 above)
Comment (santhnm2) [reply megatron/core/inference/moe/vllm_fused_moe.py]: Done

File: megatron/core/transformer/moe/moe_layer.py (hunk 5 above)
Comment (santhnm2) [reply megatron/core/transformer/moe/moe_layer.py]: Move check to top of file

File: megatron/core/transformer/moe/token_dispatcher_inference.py (hunk 6 above)
Comment (santhnm2) [reply megatron/core/transformer/moe/token_dispatcher_inference.py]: Fixed

**Review by kvareddy — APPROVED** [review]

**Review by Victarry — APPROVED** [review]

**Review by ericharper — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]

**Review by sidsingh-nvidia — APPROVED** [review]
